packages/transactions/src/builders.ts

```diff
@@ -1692,3 +1723,70 @@ export async function estimateTransactionFeeWithFallback(
     throw error;
   }
 }
+
+/** @internal multi-sig signing re-use */
+function mutatingSignAppendMultiSig(
+  /** **Warning:** method mutates `transaction` */
+  transaction: StacksTransaction,
+  publicKeys: string[],
+  signerKeys: string[],
+  address?: string
+) {
+  if (isSingleSig(transaction.auth.spendingCondition)) {
+    throw new Error('Transaction is not a multi-sig transaction');
+  }
+
+  const signer = new TransactionSigner(transaction);
+
+  const pubs = ensureValidMultiSigPublicKeyOrder(
+    publicKeys,
+    transaction.auth.spendingCondition.signaturesRequired,
+    transaction.auth.spendingCondition.hashMode,
+    address ? createAddress(address).hash160 : undefined
+  );
+
+  // sign in order of public keys
+  for (const publicKey of pubs) {
+    const signerKey = signerKeys.find(key => bytesToHex(pubKeyfromPrivKey(key).data) === publicKey);
+    if (signerKey) {
+      // either sign and append message signature (which allows for recovering the public key)
+      signer.signOrigin(createStacksPrivateKey(signerKey));
+    } else {
+      // or append the public key (which did not sign here)
+      signer.appendOrigin(publicKeyFromBytes(hexToBytes(publicKey)));
+    }
+  }
+}
+
+/** @internal Get the matching public-keys array for a multi-sig address */
+function ensureValidMultiSigPublicKeyOrder(
```

Comment:
Using the word "ensure" implies this function either returns a boolean, or modifies it's arguments. Instead, it returns the pubkeys in the correct order. I would rename this to something like `sortPublicKeysForAddress()`

Reply:
We typically have our ensures `throw` if the bad condition occurs, but this is internal so idc either way 👍

Reply:
Makes sense, but could also return a `null` and let the calling function decide whether to `throw` or recover. Seems like the function is trying to do to things that maybe shouldn't go together